SB-1.md

```diff
@@ -0,0 +1,76 @@
+# SB-1 (AMQP Backed Implementation)
+
+**Status**: Under Review
+
+This proposal introduces a method of using a message queue (Anything implementing
+AMQP 0.9.1, focusing on RabbitMQ) rather than the internal, custom queue.
+
+## Context
+
+Much of Seabird core has been built around a custom message queue, limited to a
+single node. This has made more complex operations harder to implement and led
+to less robust code.
+
+## Goals
+
+- Remove custom internal message queue
+- Replace as much logic as possible with queue routing
+
+## Potential Problems
+
+- This loses information about the current streams, such as channels,
+  and backend info. This may be acceptable for now and is related to the API refactor in [SB-3](./SB-3.md)
+
+## Details
+
+There are two types of "requests" in Seabird: request-response and streaming.
+Both of these will need to be backed by a message queue. All of this will be
+handled by seabird-core, so the actual implementation of clients should not
+change.
+
+These request types refer to the internal implementation of seabird-core.
+
+Queues will come in three forms:
+
+- `seabird.reply.*`, a separate queue per RPC request, an exclusive, direct
+  queue. This could potentially have a name provided by rabbit.
+- `seabird.command.*`, a separate queue per backend connection, an exclusive,
+  direct queue. The usage will be described under request-response.
+- `seabird.events`, a single "firehose", header exchange. All backend events will
+  be submitted to this queue. On startup, seabird-core will attempt to create
+  this queue.
```

Comment:
Please add more details about which messages would be included in each of these. I feel like we're missing a lot of detail.

It would be really nice if we could have something like

```
seabird.message  // all raw messages
    seabird.message.{backend}  // all raw messages from {backend}
seabird.command  // all commands
    seabird.command.{backend}  // all commands from {backend}
```

That way we could pick and choose what we're subscribed to. What kind of overhead would that mean?

Reply:
Yeah, some of this was originally meant as a way for me to brain dump and make sure I'm not doing anything that wouldn't work. I can flesh these out.

I haven't really thought through how to separate out messages just yet, but making sure subscriptions can be useful will definitely be part of that.